File: bin/process_rinex_requests.php
<?php
declare(strict_types=1);

/**
 * Обработка одного "готового" фонового запроса RINEX-данных (см. rinex.php
 * — там пользователь описывает критерии и создаёт запись в rinex_requests
 * со статусом 'pending'; эта запись становится доступна для скачивания на
 * странице "Готовые данные" — rinex_requests.php).
 *
 * За один запуск обрабатывается ОДИН запрос (тот, что ждёт дольше всех) —
 * простая модель без отдельного демона: запускать по расписанию
 * (Планировщик Windows) часто, например раз в 1-2 минуты, как и остальные
 * bin/*.php скрипты в проекте.
 *
 * Запускать: php C:\path\to\bin\process_rinex_requests.php
 */

require __DIR__ . '/../app/lib/db.php';
require __DIR__ . '/../app/lib/cli.php';
require __DIR__ . '/../app/lib/gnss_ftp.php';
require __DIR__ . '/../app/lib/rinex_merge.php';
require_cli_or_token();

function log_line(string $msg): void
{
    cli_out('[' . date('Y-m-d H:i:s') . '] ' . $msg);
}

// Разрыв между концом одного файла и началом следующего, при котором они
// СЧИТАЮТСЯ "идущими подряд" и склеиваются в один — больше этого считается
// дыркой (станция не присылала данные/обрыв сессии), из неё получаются
// два отдельных файла вместо одного с пропуском посередине.
const RINEX_GAP_TOLERANCE_MINUTES = 5;

$resultsDir = __DIR__ . '/../uploads/rinex_results';
if (!is_dir($resultsDir)) {
    mkdir($resultsDir, 0755, true);
}

$pdo = db();

// Атомарный захват одного запроса — SELECT...FOR UPDATE SKIP LOCKED внутри
// транзакции защищает от двух параллельных запусков скрипта (если предыдущий
// запуск ещё не закончился к моменту следующего по расписанию) от обработки
// одного и того же запроса дважды.
$pdo->beginTransaction();
$claim = $pdo->query(
    "SELECT id FROM rinex_requests WHERE status = 'pending' ORDER BY created_at LIMIT 1 FOR UPDATE SKIP LOCKED"
)->fetch();
if (!$claim) {
    $pdo->rollBack();
    log_line('Нет ожидающих запросов');
    exit;
}
$pdo->prepare("UPDATE rinex_requests SET status = 'processing', started_at = NOW() WHERE id = :id")
    ->execute(['id' => $claim['id']]);
$pdo->commit();

$requestId = (int)$claim['id'];
$stmt = $pdo->prepare('SELECT * FROM rinex_requests WHERE id = :id');
$stmt->execute(['id' => $requestId]);
$request = $stmt->fetch();

set_time_limit(0); // сборка за широкий диапазон дат/много станций может занять заметно больше 120с по умолчанию

function fail_request(PDO $pdo, int $id, string $message): void
{
    $pdo->prepare("UPDATE rinex_requests SET status = 'error', error_message = :msg, completed_at = NOW() WHERE id = :id")
        ->execute(['id' => $id, 'msg' => substr($message, 0, 500)]);
    log_line("Запрос #$id: ОШИБКА — $message");
}

try {
    $stations = array_values(array_filter(array_map('trim', explode(',', $request['stations']))));
    // Явно UTC — значения в БД уже UTC (см. rinex.php), но PHP без явной
    // зоны интерпретирует строку в date.timezone из php.ini, который в
    // проекте никогда не настраивается явно (зависит от хостинга).
    $dateFrom = new DateTimeImmutable($request['date_from_utc'], new DateTimeZone('UTC'));
    $dateTo = new DateTimeImmutable($request['date_to_utc'], new DateTimeZone('UTC'));
    $wantTypes = [];
    if ((int)$request['want_obs'] === 1) $wantTypes['MO'] = true;
    if ((int)$request['want_nav'] === 1) $wantTypes['MN'] = true;
    $mergeByDay = (int)$request['merge_by_day'] === 1;

    if (!$stations || !$wantTypes) {
        fail_request($pdo, $requestId, 'Не указаны станции или типы файлов');
        exit;
    }

    $conn = gnss_ftp_connect();
    if ($conn === false) {
        fail_request($pdo, $requestId, 'Не удалось подключиться к FTP gnss.host');
        exit;
    }

    $tmpDir = $resultsDir . '/tmp_' . $requestId;
    $finalDir = $resultsDir . '/final_' . $requestId;
    mkdir($tmpDir, 0755, true);
    mkdir($finalDir, 0755, true);

    // Шаг 1 — скачиваем с FTP все файлы, попадающие в [dateFrom, dateTo] по
    // их СОБСТВЕННОЙ метке времени (gnss_parse_file_timestamp), а не по
    // границам папки дня — последняя папка дня содержит часы и до, и после
    // запрошенного окна, фильтруем точно.
    $downloaded = []; // [station][type] => [ ['local_path'=>..., 'timestamp'=>DateTimeImmutable, 'period_minutes'=>?int], ... ]
    $cursorDay = $dateFrom->setTime(0, 0);
    $lastDay = $dateTo->setTime(0, 0);
    $fileIndex = 0;
    while ($cursorDay <= $lastDay) {
        $dayFolder = gnss_day_folder($cursorDay);
        foreach ($stations as $station) {
            $files = gnss_ftp_list_files($conn, $dayFolder, $station);
            foreach ($files as $f) {
                $ts = gnss_parse_file_timestamp($f['name']);
                if ($ts === null || $ts < $dateFrom || $ts > $dateTo) {
                    continue;
                }
                $type = stripos($f['name'], '_MO.') !== false ? 'MO'
                    : (stripos($f['name'], '_MN.') !== false ? 'MN' : null);
                if ($type === null || !isset($wantTypes[$type])) {
                    continue;
                }
                $fileIndex++;
                $localPath = $tmpDir . '/dl_' . $fileIndex;
                if (!@ftp_get($conn, $localPath, '/' . $f['path'], FTP_BINARY)) {
                    continue; // пропускаем недостающий/неудачно скачанный файл, не валим весь запрос
                }
                $downloaded[$station][$type][] = [
                    'local_path' => $localPath,
                    'timestamp' => $ts,
                    'period_minutes' => gnss_parse_file_period_minutes($f['name']),
                ];
            }
        }
        $cursorDay = $cursorDay->modify('+1 day');
    }
    ftp_close($conn);

    if (!$downloaded) {
        fail_request($pdo, $requestId, 'Не найдено файлов по заданным станциям/датам/типам');
        // Чистим пустые временные папки.
        @rmdir($tmpDir);
        @rmdir($finalDir);
        exit;
    }

    // Шаг 2 — группировка/склейка. Без merge_by_day — каждый скачанный файл
    // просто переносится в финальную папку под читаемым именем (без склейки
    // вообще, на случай если пользователь явно не просил объединять часы).
    $fileCount = 0;
    foreach ($downloaded as $station => $byType) {
        foreach ($byType as $type => $files) {
            if (!$mergeByDay) {
                foreach ($files as $f) {
                    $name = $station . '_' . $f['timestamp']->format('Ymd_Hi') . '_' . $type . '.rnx';
                    rename($f['local_path'], $finalDir . '/' . $name);
                    $fileCount++;
                }
                continue;
            }
            $groups = rinex_group_contiguous_files($files, RINEX_GAP_TOLERANCE_MINUTES);
            foreach ($groups as $group) {
                $first = $group[0]['timestamp'];
                $last = end($group)['timestamp'];
                $name = $station . '_' . $type . '_' . $first->format('Ymd') . '_'
                    . $first->format('Hi') . '-' . $last->format('Hi') . '.rnx';
                $localPaths = array_map(fn($g) => $g['local_path'], $group);
                if (count($localPaths) === 1) {
                    rename($localPaths[0], $finalDir . '/' . $name);
                } else {
                    rinex_merge_group($localPaths, $finalDir . '/' . $name);
                }
                $fileCount++;
            }
        }
    }

    // dl_* файлы нужны были только до этого момента (либо переименованы,
    // либо уже прочитаны при склейке) — что осталось (входило в склеенные
    // группы из >1 файла) убираем.
    foreach (glob($tmpDir . '/dl_*') as $leftover) {
        @unlink($leftover);
    }
    @rmdir($tmpDir);

    // Шаг 3 — архивируем финальную папку в один .zip для скачивания.
    $zipRelPath = $requestId . '.zip';
    $zipPath = $resultsDir . '/' . $zipRelPath;
    $zip = new ZipArchive();
    $zip->open($zipPath, ZipArchive::CREATE | ZipArchive::OVERWRITE);
    foreach (glob($finalDir . '/*') as $f) {
        $zip->addFile($f, basename($f));
    }
    $zip->close();

    foreach (glob($finalDir . '/*') as $f) {
        @unlink($f);
    }
    @rmdir($finalDir);

    $pdo->prepare(
        "UPDATE rinex_requests SET status = 'done', result_path = :path, file_count = :count, completed_at = NOW() WHERE id = :id"
    )->execute(['path' => $zipRelPath, 'count' => $fileCount, 'id' => $requestId]);

    log_line("Запрос #$requestId: готово, файлов в архиве: $fileCount");
} catch (Throwable $e) {
    fail_request($pdo, $requestId, $e->getMessage());
}

File: pyserver/rinex/management/commands/process_rinex_requests.py
"""
Port of bin/process_rinex_requests.php. Processes ONE "pending" background
RINEX request per run (the one that''s been waiting longest) -- simple
model without a separate daemon: run on a schedule (systemd timer, every
1-2 minutes), same cadence as the PHP original under Task Scheduler.
"""

import shutil
import zipfile
from datetime import timedelta
from pathlib import Path

from django.conf import settings
from django.core.management.base import BaseCommand
from django.db import transaction
from django.utils import timezone

from rinex.gnss_ftp import gnss_day_folder, gnss_ftp_connect, gnss_ftp_list_files, gnss_parse_file_period_minutes, gnss_parse_file_timestamp
from rinex.models import RinexRequest
from rinex.rinex_merge import rinex_group_contiguous_files, rinex_merge_group

# Gap between the end of one file and the start of the next at which they
# still count as "back to back" and get merged into one -- more than this
# is a hole (station didn't send data / session drop), producing two
# separate files instead of one with a gap in the middle.
RINEX_GAP_TOLERANCE_MINUTES = 5


class Command(BaseCommand):
    help = "Process one pending RINEX background request (FTP fetch + optional per-day merge + zip)."

    def handle(self, *args, **options):
        results_dir = Path(settings.UPLOADS_ROOT) / "rinex_results"
        results_dir.mkdir(parents=True, exist_ok=True)

        with transaction.atomic():
            claim = (
                RinexRequest.objects.select_for_update(skip_locked=True)
                .filter(status="pending").order_by("created_at").first()
            )
            if not claim:
                self.stdout.write("Нет ожидающих запросов")
                return
            claim.status = "processing"
            claim.started_at = timezone.now()
            claim.save(update_fields=["status", "started_at"])
            request_id = claim.id

        rr = RinexRequest.objects.get(id=request_id)

        try:
            stations = [s.strip() for s in rr.stations.split(",") if s.strip()]
            date_from = rr.date_from_utc
            date_to = rr.date_to_utc
            want_types = set()
            if rr.want_obs:
                want_types.add("MO")
            if rr.want_nav:
                want_types.add("MN")
            merge_by_day = bool(rr.merge_by_day)

            if not stations or not want_types:
                self._fail(rr, "Не указаны станции или типы файлов")
                return

            ftp = gnss_ftp_connect()
            if ftp is None:
                self._fail(rr, "Не удалось подключиться к FTP gnss.host")
                return

            tmp_dir = results_dir / f"tmp_{request_id}"
            final_dir = results_dir / f"final_{request_id}"
            tmp_dir.mkdir(parents=True, exist_ok=True)
            final_dir.mkdir(parents=True, exist_ok=True)

            # Step 1 -- download every FTP file falling in [date_from, date_to]
            # by its OWN timestamp (gnss_parse_file_timestamp), not by the day
            # folder's bounds -- the last day's folder holds hours before and
            # after the requested window too, filter precisely.
            downloaded: dict[str, dict[str, list]] = {}
            cursor_day = date_from.replace(hour=0, minute=0, second=0, microsecond=0)
            last_day = date_to.replace(hour=0, minute=0, second=0, microsecond=0)
            file_index = 0
            try:
                while cursor_day <= last_day:
                    day_folder = gnss_day_folder(cursor_day)
                    for station in stations:
                        for f in gnss_ftp_list_files(ftp, day_folder, station):
                            ts = gnss_parse_file_timestamp(f["name"])
                            if ts is None or ts < date_from or ts > date_to:
                                continue
                            name_lower = f["name"].lower()
                            file_type = "MO" if "_mo." in name_lower else ("MN" if "_mn." in name_lower else None)
                            if file_type is None or file_type not in want_types:
                                continue
                            file_index += 1
                            local_path = tmp_dir / f"dl_{file_index}"
                            try:
                                with open(local_path, "wb") as fh:
                                    ftp.retrbinary(f"RETR /{f['path']}", fh.write)
                            except OSError:
                                continue  # skip missing/failed file, don't fail the whole request
                            downloaded.setdefault(station, {}).setdefault(file_type, []).append({
                                "local_path": str(local_path), "timestamp": ts,
                                "period_minutes": gnss_parse_file_period_minutes(f["name"]),
                            })
                    cursor_day += timedelta(days=1)
            finally:
                ftp.close()

            if not downloaded:
                self._fail(rr, "Не найдено файлов по заданным станциям/датам/типам")
                tmp_dir.rmdir()
                final_dir.rmdir()
                return

            # Step 2 -- group/merge. Without merge_by_day, each downloaded file
            # is just moved into the final folder under a readable name.
            file_count = 0
            for station, by_type in downloaded.items():
                for file_type, files in by_type.items():
                    if not merge_by_day:
                        for f in files:
                            name = f"{station}_{f['timestamp'].strftime('%Y%m%d_%H%M')}_{file_type}.rnx"
                            shutil.move(f["local_path"], final_dir / name)
                            file_count += 1
                        continue
                    groups = rinex_group_contiguous_files(files, RINEX_GAP_TOLERANCE_MINUTES)
                    for group in groups:
                        first = group[0]["timestamp"]
                        last = group[-1]["timestamp"]
                        name = (
                            f"{station}_{file_type}_{first.strftime('%Y%m%d')}_"
                            f"{first.strftime('%H%M')}-{last.strftime('%H%M')}.rnx"
                        )
                        local_paths = [g["local_path"] for g in group]
                        if len(local_paths) == 1:
                            shutil.move(local_paths[0], final_dir / name)
                        else:
                            rinex_merge_group(local_paths, str(final_dir / name))
                        file_count += 1

            # dl_* files were only needed up to this point (either renamed or
            # already read during merging) -- clean up whatever's left (files
            # that went into groups of >1).
            for leftover in tmp_dir.glob("dl_*"):
                leftover.unlink(missing_ok=True)
            tmp_dir.rmdir()

            # Step 3 -- zip the final folder into one downloadable archive.
            zip_rel_path = f"{request_id}.zip"
            zip_path = results_dir / zip_rel_path
            with zipfile.ZipFile(zip_path, "w") as zf:
                for f in final_dir.glob("*"):
                    zf.write(f, f.name)

            for f in final_dir.glob("*"):
                f.unlink(missing_ok=True)
            final_dir.rmdir()

            rr.status = "done"
            rr.result_path = zip_rel_path
            rr.file_count = file_count
            rr.completed_at = timezone.now()
            rr.save(update_fields=["status", "result_path", "file_count", "completed_at"])
            self.stdout.write(f"Запрос #{request_id}: готово, файлов в архиве: {file_count}")
        except Exception as exc:
            self._fail(rr, str(exc))

    def _fail(self, rr: RinexRequest, message: str):
        rr.status = "error"
        rr.error_message = message[:500]
        rr.completed_at = timezone.now()
        rr.save(update_fields=["status", "error_message", "completed_at"])
        self.stdout.write(f"Запрос #{rr.id}: ОШИБКА — {message}")